packages/opentelemetry-exporter-prometheus/src/prometheus.ts

```diff
@@ -0,0 +1,305 @@
+    const metricName = this._getPrometheusMetricName(readableMetric.descriptor);
+    const metric = this._registry.getSingleMetric(metricName);
+
+    /**
+     * Prometheus library does aggregation, which means its inc method must be called with
+     * the value to be incremented by. It does not have a set method. As our ReadableMetric
+     * contains the current value, not the value to be incremented by, we destroy and
+     * recreate counters when they are updated.
+     *
+     * This works because counters are identified by their name and no other internal ID
+     * https://prometheus.io/docs/instrumenting/exposition_formats/
+     */
+    if (metric instanceof Counter) {
+      this._registry.removeSingleMetric(metricName);
+    } else if (metric) return metric;
+
+    return this._newMetric(readableMetric, metricName);
+  }
+
+  private _newMetric(
+    readableMetric: ReadableMetric,
+    name: string
+  ): Metric | undefined {
+    const metricObject = {
+      name,
+      // prom-client throws with empty description which is our default
+      help: readableMetric.descriptor.description || 'description missing',
+      labelNames: readableMetric.descriptor.labelKeys,
+      // list of registries to register the newly created metric
+      registers: [this._registry],
+    };
+
+    switch (readableMetric.descriptor.type) {
+      case MetricDescriptorType.COUNTER_DOUBLE:
+      case MetricDescriptorType.COUNTER_INT64:
+        return new Counter(metricObject);
+      case MetricDescriptorType.GAUGE_DOUBLE:
+      case MetricDescriptorType.GAUGE_INT64:
+        return new Gauge(metricObject);
+      default:
+        // Other metric types are currently unimplemented
+        return undefined;
+    }
+  }
+
+  private _getPrometheusMetricName(descriptor: MetricDescriptor): string {
+    return this._sanitizePrometheusMetricName(
+      this._prefix ? `${this._prefix}_${descriptor.name}` : descriptor.name
+    );
+  }
+
+  /**
+   * Ensures metric names are valid Prometheus metric names by removing
+   * characters allowed by OpenTelemetry but disallowed by Prometheus.
+   *
+   * https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+   *
+   * 1. Names must match `[a-zA-Z_:][a-zA-Z0-9_:]*`
+   *
+   * 2. Colons are reserved for user defined recording rules.
+   * They should not be used by exporters or direct instrumentation.
+   *
+   * OpenTelemetry metric names are already validated in the Meter when they are created,
+   * and they match the format `[a-zA-Z][a-zA-Z0-9_.\-]*` which is very close to a valid
+   * prometheus metric name, so we only need to strip characters valid in OpenTelemetry
+   * but not valid in prometheus and replace them with '_'.
+   *
+   * @param name name to be sanitized
+   */
+  private _sanitizePrometheusMetricName(name: string): string {
+    return name.replace(this._invalidCharacterRegex, '_'); // replace all invalid characters with '_'
```

Comment:
Is it possible this could cause duplicates? Should we write a warning in that case (could be a follow up task)?

Reply:
We could replace period with __DOT__ or similar